dynamo-adapter.js

```diff
@@ -0,0 +1,126 @@
+const {promisify} = require('util');
+const {DynamoTable} = require('./slim');
+
+const _isCallback = (fn) => typeof fn === 'function';
+
+const _validateCall = (methodName, args, minLength) => {
+  if (args.length < minLength)
+    throw new Error(`Incorrect args supplied to method ${methodName}: expected: ${minLength} got: ${args.length}`);
+}
+
+/**
+  Provides an adapter for the DynamoTable class to allow for promisified
+  api usage when omitting the cb parameter. Does not use ES6 classes
+  because some dependents use assign to copy the instance which is only a
+  shallow copy and insufficient to clone an ES6 class instance.
+*/
+function DynamoAdapter(name, options) {
+  this._delegate = new DynamoTable(name, options);
+
+  // Delegate externally accessed properties
+  Object.keys(this._delegate).forEach(key => {
+    this[key] = this._delegate[key];
+  });
+
+  // NOTE: Functions are defined in the constructor so they can be rebound
+  // later to allow for passing methods without the instance.
+  this._call = function(methodName, minLength, ...args) {
```

Comment:
Had to move all these functions into the constructor so I can rebind them to the instance. Later some are used without the instance such as in `async` functions which worked previously because of some monkey patching over the method where it no longer used any instance attributes, now everything needs at least `this._call` and `this._delegate` so such cases failed. e.g.

```coffeescript
# This would fail with the adapter as this._call is not defined
async.each items, tables.thing.addNew, cb
```

Reply:
how about creating once as prototype and using a loop to bind them in the constructor? Just kind of look more conventional I guess, not sure about the perf difference though.

Reply:
also with the assignments if we prob can avoid repeating `function(`
```
Object.assign(this, {
   copy() {
      //..
   addNew() {
```

Reply:
there are a few places where we use `async.each items, tables.thing.addNew.bind(tables.thing), cb` to get around this, I guess that will not break and does not need to be updated?

Reply:
@sonhanguyen When they were on the prototype they dont show in `Object.keys` probably would be there with `Object.getOwnPropertyNames` could change.

@sonhanguyen Originally did the `function` statements due to some external code using assign which doesn't copy some functions when on es6 class and when on proto needs `this` context. But I added the copy function so possibly I can even go back to es6 class

@vinteo hmm not sure will have a look, definitely that block you posted should be fine, I did find some without the binding just for `addNew` cause it was patched on, its probable your example is found on the other methods though which need the bind else they don't work so will check up on them

Reply:
Changed per my above to es6 should be ok just finding some minor issues where the stuff is hacked up to resolve externally.